### PR TITLE
Improvements for policy creation at permissions screen

### LINKED ISSUE #20884: [Admin Console v2] Policy creation at Permissions screen missing
### Description

Console V1:
1. Policy can be created on the fly
2. User can navigate to the policy

**Console v2 v21.1.1 is missing these features:**
image

**In console V1 they were available:**
image

### Discussion

_No response_

### Motivation

With console v2 a new policy has to be created beforehand at realm-management client, while it could just be created at the permissions screen before at console v1.

For reviewing a Permission it should be possible to navigate to the Policy which is selected for the Permission.

### Details

Feature proposal: 

- Allow to click on Policy to navigate to it
- Add "Create Policy" button or an option in the Policy selection dropdown
image

### PR DESCRIPTION
**Pull request by andreas-blaettlinger**
Allow navigation to policy on permission screen

Closes #20884

### CODE REVIEW

**Comment by andreas-blaettlinger** [comment]
I'm still looking for a solution for feature proposal#2: "Add "Create Policy" button or an option in the Policy selection dropdown"

**Comment by jonkoops** [comment]
@andreas-blaettlinger perhaps it would be best to land this separately, and focus on landing this one for now?

**Comment by andreas-blaettlinger** [comment]
@jonkoops  I agree, lets focus on the navigation first. Also fixed the failing cypress test.

**Comment by jonkoops** [comment]
The code looks alright to me, @xianli123 can I ask you to do a UX review on this?

**Comment by xianli123** [comment]
Thanks @andreas-blaettlinger @jonkoops I am not sure why we want to navigate to the policy's page. Could you please explain the use case?

**Comment by andreas-blaettlinger** [comment]
@xianli123 @jonkoops In the past it was also possible with the old UI. Now in the new UI you have to jump a lot around multiple (not so easy to find) pages, we would like to make this easier, kinda similar to the old UI. For more information you can check #20884.

**Comment by xianli123** [comment]
The implementation of this PR can fit thescenarios well. But there are 2 problems.

1. The options in the dropdown should not be a link. The option can't have interactive children. Assistive technologies, such as screen readers, might ignore or respond unexpectedly to such nested controls.
2. Once users leave the permission creation page for the policy page, users will not return to the permission creation page and all unsaved settings will be cleared.

**Comment by xianli123** [comment]
Thanks @andreas-blaettlinger , Does the user want to go policy's details page because the user wants to know more details about one specific policy?

**Comment by andreas-blaettlinger** [comment]
> The implementation of this PR can fit thescenarios well. But there are 2 problems.
> 
>     1. The options in the dropdown should not be a link. The option can't have interactive children. Assistive technologies, such as screen readers, might ignore or respond unexpectedly to such nested controls.
> 
>     2. Once users leave the permission creation page for the policy page, users will not return to the permission creation page and all unsaved settings will be cleared.

I agree on your first point and will look for a solution.

To your second point:

That's exactly the same behaviour as in the old UI. IMO the PR handles is neither better nor worse than before. You could still use the back-function of your browser to at least navigate back and forth; but keeping the unsafed settings while navigating is a bigger thing to fix/implement. First of all we would have to decide how we want to implement this (e.g. with url parameters, ...). I think we could make such fundamental improvements at a later point, wdyt @xianli123 ?

> Thanks @andreas-blaettlinger , Does the user want to go policy's details page because the user wants to know more details about one specific policy?

Yes exactly, for every existing policy we want to allow the user to navigate to the details of each policy.

**Comment by xianli123** [comment]
@andreas-blaettlinger Sorry for the late reply. 

> for every existing policy we want to allow the user to navigate to the details of each policy.

Based on your explanation, we may provide a way to display some policy details on the permission creation page instead of navigating to the policy's details page. In this way, users don't need to leave the permission creation page. It will not break the creation flow. What do you think?

**Comment by andreas-blaettlinger** [comment]
@xianli123 After brainstorming we came up with an idea. We would really like to keep the links but we agree that we do not want to break the flow (accidentally). What if a user presses the link and then it checks if there are unsaved changes. If so, we show this confirmation dialog and make the user aware of it. If he really wants to navigate there, he just confirms the dialog and is redirected, otherwise he cancels and nothing happens. Could look something like this when clicking on a link.

image

WDYT?

**Comment by xianli123** [comment]
@andreas-blaettlinger Sorry for the late reply. Thanks for your good suggestions. Based on your requirements, I think users just want to know some details of the policy, so I think we can add a function to help users review the details of the policy on the permission creation page.

I came up with a solution as shown below. When hovering over the policy, there will be a tooltip popup as shown in the first screenshot. The tooltip displays some basic info of the policy. If users want to view more details, there is a "More details" link. After clicking the "More details" link, the details modal will pop up as shown in the second screenshot. In this way, it's not necessary to leave the current page. What do you think of this solution? Thanks!

**Tooltip**

> Screenshot 2023-07-10 at 15 11 47

**Details modal**

> Screenshot 2023-07-10 at 15 12 18

**Comment by jbman** [comment]
Hi @xianli123, 
it is more valuable to actually navigate to the policy instead of creating different views on the data. Users will get one consistent representation they can learn and recognize. They can go to a policy and decide to read only or change it. This seems to be overall concept of the admin console: Having data editable right away. I think this concept works quite good.

But whats definitely missing is returning from the permission into the policy. However back navigation is already broken when going from e.g "Users / Permissions" tab into a Permission.

Clicking on cancel at the Permission end up at the realm management client with the list of Permissions:


To sum it up: Navigation into a Policy from the Permissions view [as suggested by @andreas-blaettlinger ](https://github.com/keycloak/keycloak/pull/21411#issuecomment-1625329333) is already valuable. Would like to get that into Keycloak as soon as possible for customers because they miss it when comparing against the previous admin console.
I would like to start a separate discussion so that back-navigation will work when navigating in policies. 

xianli123, could we move on like this?

**Comment by jbman** [comment]
Hi @xianli123, 
it is more valuable to actually navigate to the policy instead of creating different views on the data. Users will get one consistent representation they can learn and recognize. They can go to a policy and decide to read only or change it. This seems to be overall concept of the admin console: Having data editable right away. I think this concept works quite good.

But whats definitely missing is returning from the permission into the policy. However back navigation is already broken when going from e.g "Users / Permissions" tab into a Permission.

Clicking on cancel at the Permission end up at the realm management client with the list of Permissions:


To sum it up: Navigation into a Policy from the Permissions view [as suggested by @andreas-blaettlinger ](https://github.com/keycloak/keycloak/pull/21411#issuecomment-1625329333) is already valuable. Would like to get that into Keycloak as soon as possible for customers because they miss it when comparing against the previous admin console.
I would like to start a separate discussion so that back-navigation will work when navigating in policies. 

xianli123, could we move on like this?

**Comment by xianli123** [comment]
Hi @jbman Thanks for your explanition. I agree with you. We can move forward like @andreas-blaettlinger suggested. The dialog is a good way to inform users. Let's move on like this. Thank you!

**Comment by jbman** [comment]
Thank you!
General navigation problems with Policies and Permissions are now summarized at discussion https://github.com/keycloak/keycloak/discussions/21547

**Comment by andreas-blaettlinger** [comment]
I pushed the latest changes, can be reviewed now. FYI @jonkoops  @xianli123

File: js/apps/admin-ui/public/locales/en/common.json [hunk 1]
```diff
@@ -215,5 +215,7 @@
   "forbidden_other": "Forbidden, permissions needed:",
   "noRealmRolesToAssign": "There are no realm roles to assign",
   "loadingRealms": "Loading realms…",
-  "customAttribute": "Custom Attribute…"
+  "customAttribute": "Custom Attribute…",
+  "unsavedChangesTitle": "Unsaved changes",
+  "unsavedChangesConfirm": "You have unsaved changes. Do you really want to leave the page?"
```
Comment (jonkoops) [inline js/apps/admin-ui/public/locales/en/common.json:220]: I am not sure these labels belong in the `common` namespace. They only seem to be used in the clients section, so perhaps it would be better to add them to `clients.json`.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 2]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
     [search]
   );
 
+  const [toggleUnsavedChangesDialog, UnsavedChangesConfirm] = useConfirmDialog({
+    titleKey: t("common:unsavedChangesTitle"),
+    messageKey: t("common:unsavedChangesConfirm"),
+    continueButtonLabel: t("common:continue"),
+    continueButtonVariant: ButtonVariant.danger,
+    onConfirm: async () => {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:136]: There is no need for this to be an `async` function, as there is no Promise returned from it.
```suggestion
    onConfirm() {
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 3]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
         {p.name}
       </SelectOption>
     ));
 
+  const toChipGroupItems = (
+    field: ControllerRenderProps<PolicyRepresentation, Type>
+  ) => {
+    return (
+      <ChipGroup>
+        {(field.value as string[]).map((permissionId) => {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:160]: Rather than type-casting this value, I'd prefer if we properly handle the possibility this value might be undefined. Trying to out-smart the compiler can lead to undefined behavior. 
```suggestion
        {field.value?.map((permissionId) => {
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 4]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+          const policy = items.find(
+            (permission) => permission.id === permissionId
+          );
+          if (policy) {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:164]: I am more of a fan of inverting this condition and using a early return. This prevents deeply nested statements, and improves readability of the code.

```ts
if (!policy) {
  return null;
}

// The rest of the code.
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 5]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+            return (
+              <Chip
+                key={policy.id}
+                onClick={(e: React.MouseEvent) => {
+                  e.stopPropagation();
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:169]: There should be no need to add the type here, it can be inferred from the signature of the prop. Also, let's avoud single letter variable names where possible.
```suggestion
                onClick={(event) => {
                  event.stopPropagation();
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 6]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+                  field.onChange(
+                    field.value?.filter((id: string) => id !== policy.id)
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:171]: Same here, the type of `id` can already be inferred.
```suggestion
                    field.value?.filter((id) => id !== policy.id)
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 7]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+                  );
+                }}
+              >
+                {policy.type ? (
+                  <Link
+                    to={""}
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:177]: For accessibility reasons a link always needs to have an `href`. So let's make sure it exists, even if navigation can be prevented.

```tsx
const to = toPolicyDetails({
  realm: realm,
  id: clientId,
  policyId: policy.id!,
  policyType: policy.type!,
});

<Link to={to}>
```

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 8]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+                    onClick={(e) => {
+                      e.stopPropagation();
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:179]: This should only be done in case `isDirty` is `true`.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 9]
```diff
@@ -108,57 +128,133 @@ export const ResourcesPolicySelect = ({
...
+                  e.stopPropagation();
+                  field.onChange(
+                    field.value?.filter((id: string) => id !== policy.id)
+                  );
+                }}
+              >
+                {policy.type ? (
+                  <Link
+                    to={""}
+                    onClick={(e) => {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:178]: Let's also use a full variable name here.
```suggestion
                    onClick={(event) => {
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by andreas-blaettlinger** [comment]
@jonkoops I followed your suggestions, better that way?

**Comment by xianli123** [comment]
@andreas-blaettlinger Great. Thanks!

**Comment by andreas-blaettlinger** [comment]
Resolved merge conflict.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 10]
```diff
@@ -108,57 +128,141 @@ export const ResourcesPolicySelect = ({
...
+                field.onChange(field.value?.filter((id) => id !== policy.id));
+              }}
+            >
+              {policy.type ? (
+                <Link
+                  to={to}
+                  onClick={(event) => {
+                    if (isDirty) {
+                      event.preventDefault();
+                      event.stopPropagation();
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:190]: Do we have to stop propagation? I believe preventing the default behaviour here should be enough, but I can be wrong.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 11]
```diff
@@ -108,57 +128,141 @@ export const ResourcesPolicySelect = ({
...
+                      setOpen(false);
+                      setClickedPolicy(policy);
+                      toggleUnsavedChangesDialog();
+                    } else {
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:194]: There should be no need for this `else` clause, as the default behavior should be that the link is followed.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 12]
```diff
@@ -108,57 +128,141 @@ export const ResourcesPolicySelect = ({
...
+        {field.value?.map((permissionId) => {
+          const policy = items.find(
+            (permission) => permission.id === permissionId
+          );
+
+          if (!policy) {
+            return null;
+          }
+
+          const to = toPolicyDetails({
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:169]: This link is also used higher up in the unsaved changes dialog. so perhaps it can be re-used there as well?

**Review by jonkoops — CHANGES_REQUESTED** [review]
There are some linting issues that need to be resolved. We recently updated some dependencies which changed the formatting, so you'll have to run `pnpm i` before to bring them up-to-date.

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx (hunk 10 above)
Comment (andreas-blaettlinger) [reply js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:190]: IMO its needed for following case:

grafik

When you click on a Chip-Link while the SelectOptions are shown, your first click would just hide the SelectOptions (even when you click on the Dialogs Continue/Cancel buttons). That's why I added a `stopPropagation()`. Preventing the default behavior just assures that the `onclick` does its thing and hinders the `<Link to={to}>` of immediately navigating somewhere to.

Comment (jonkoops) [reply js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:182]: Ahhh, yes this makes perfect sense. Consider it resolved!

File: js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx [hunk 13]
```diff
@@ -108,57 +128,124 @@ export const ResourcesPolicySelect = ({
...
 
+  const [toggleUnsavedChangesDialog, UnsavedChangesConfirm] = useConfirmDialog({
+    titleKey: t("unsavedChangesTitle"),
+    messageKey: t("unsavedChangesConfirm"),
+    continueButtonLabel: t("common:continue"),
+    continueButtonVariant: ButtonVariant.danger,
+    onConfirm: () => navigate(to(clickedPolicy!)),
+  });
+
+  const to = (policy: Policies) =>
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/authorization/ResourcesPolicySelect.tsx:139]: Good idea to put this in a function so it can be re-used :+1:

**Review by jonkoops — APPROVED** [review]
I think the code looks good now, approved :)

**Comment by jonkoops** [comment]
@xianli123 do you feel like this is ready UX-wise? If so, I'd like to merge this PR.

**Comment by xianli123** [comment]
@jonkoops Yes. The updates look good. Thanks!

**Comment by jonkoops** [comment]
Nice, thanks again @andreas-blaettlinger!
